services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/SemanticDetector.java

```diff
@@ -0,0 +1,438 @@
+              String.format(
+                  "Field %s: Double value %s is not the default value 0.0 or %s",
+                  formatFieldName(name),
+                  object,
+                  defaultValue));
+        }
+        break;
+      case BOOLEAN:
+        if (!((object instanceof Boolean) && !(boolean) object)) {
+          throw new VeniceProtocolException(
+              String.format(
+                  "Field %s: Boolean value %s is not the default value false or %s",
+                  formatFieldName(name),
+                  object,
+                  defaultValue));
+        }
+        break;
+      case STRING:
+        if (!((object instanceof String) && ((String) object).isEmpty())) {
+          throw new VeniceProtocolException(
+              String.format(
+                  "Field %s: String value %s is not the default value \"\" or %s",
+                  formatFieldName(name),
+                  object,
+                  defaultValue));
+        }
+        break;
+      case BYTES:
+        if (!((object instanceof byte[]) && ((byte[]) object).length == 0)) {
+          throw new VeniceProtocolException(
+              String.format(
+                  "Field %s: Bytes value %s is not the default value [] or %s",
+                  formatFieldName(name),
+                  object,
+                  defaultValue));
+        }
+        break;
+      default:
+        throw new VeniceProtocolException(
+            String.format(
+                "Field %s: Value %s doesn't match default value %s",
+                formatFieldName(name),
+                object,
+                defaultValue));
+    }
+  }
+
+  public static void validateEnum(Object enumValue, Schema currentSchema, Schema targetSchema, String name) {
+    if (!(enumValue instanceof String)) {
+      throw new VeniceProtocolException(
+          String.format("Field %s: Enum value %s is not a string", formatFieldName(name), enumValue));
+    }
+
+    List<String> prevEnumSymbols = currentSchema.getEnumSymbols();
+    List<String> targetEnumSymbols = targetSchema.getEnumSymbols();
+    // If the value is not in the previous enum symbols but in the target enum symbols, it is a new enum value
+    if (!prevEnumSymbols.contains(enumValue) && targetEnumSymbols.contains(enumValue)) {
+      throw new VeniceProtocolException(
+          String.format(
+              "Field %s: Enum value %s is not in the previous enum symbols but in the target enum symbols",
+              formatFieldName(name),
+              enumValue));
+    }
+  }
+
+  private static String buildFieldPath(String parent, String field) {
+    return parent.isEmpty() ? field : parent + "_" + field;
+  }
+
+  /**
+   * Format the field name to replace "_" with ".". Reason is that we use "_" as a delimiter in the field name since "." is
+   * invalid character in field name.
+   */
+  private static String formatFieldName(String fieldName) {
+    return fieldName.replace("_", ".");
+  }
```

Comment:
Just curious, why not use "." at the beginning?

Reply:
Good catch, before this change, I need to feed this fieldName into Schema.Field, which is not accepting `.` so I need to switch to `_`. Now we don't need to make it into a field anymore, I can use dot directly.